### PR TITLE
Added min/max methods for biases 

### PR DESCRIPTION
**Pull request by hhtong**

### CODE REVIEW

File: dimod/bqm/utils.pyx [hunk 1]
```diff
@@ -40,6 +42,53 @@ cdef object as_numpy_scalar(double a, np.dtype dtype):
     """Note that the memory is interpreted to match dtype, not a cast"""
     return PyArray_Scalar(&a, dtype, None)
 
+def cylmin(cyBQM bqm):
+    cdef double min_ = sys.maxsize
```
[the remark this answers is not in the document]
Comment (randomir) [reply dimod/bqm/utils.pyx:46]: Same for `cyqmin()`.

File: dimod/bqm/utils.pyx [hunk 2]
```diff
@@ -40,6 +42,53 @@ cdef object as_numpy_scalar(double a, np.dtype dtype):
...
+    cdef Py_ssize_t vi
+    for vi in range(bqm.bqm_.num_variables()):
+        val = bqm.bqm_.get_linear(vi)
+        if (val < min_):
+            min_ = val
+    
+    return min_
+
+def cylmax(cyBQM bqm):
+    cdef double max_ = 0
```
[the remark this answers is not in the document]
Comment (randomir) [reply dimod/bqm/utils.pyx:56]: Same for `cyqmax()`.

File: dimod/core/bqm.py [hunk 3]
```diff
@@ -141,6 +143,22 @@ def __setitem__(self, v, bias):
         # inherits its ability to reshape the bqm from the `.set_linear` method
         self._bqm.set_linear(v, bias)
 
+    def min(self):
+        try:
+            return cylmin(self._bqm)
+        except TypeError:
+            pass
```
Comment (randomir) [inline dimod/core/bqm.py:150]: Why would this fail?

File: tests/test_minmax.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+#
+# ================================================================================================
+
+import unittest
+import dimod
+import numpy as np
+
+num_vars = 1000
+
+class TestMinMax(unittest.TestCase):
```
Comment (randomir) [inline tests/test_minmax.py:23]: I would probably add these tests next to BQM methods tests in [`test_bqm.py`](https://github.com/dwavesystems/dimod/blob/master/tests/test_bqm.py) or [`test_binary_quadratic_model.py`](https://github.com/dwavesystems/dimod/blob/master/tests/test_binary_quadratic_model.py).

File: tests/test_minmax.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+        self.linear_minmax(bqm) 
+        self.quad_minmax(bqm)
+
+    def test_adjmap_minmax(self):
+        D = np.arange(num_vars*num_vars).reshape((num_vars, num_vars))
+        bqm = dimod.AdjMapBQM(D, 'SPIN') 
+
+        self.linear_minmax(bqm) 
+        self.quad_minmax(bqm)
```
Comment (randomir) [inline tests/test_minmax.py:65]: These can be parametrized, see for example [`test_bqm.py#L34-L47`](https://github.com/dwavesystems/dimod/blob/d697266e41f2e073c801221b7da46455a0ef1116/tests/test_bqm.py#L34-L47).

File: dimod/core/bqm.py (hunk 3 above)
Comment (arcondello) [reply dimod/core/bqm.py:150]: `cylin` and others are defined only for the cyBQMs. So this will fail for AdjDict for instance.

File: dimod/bqm/utils.pyx (hunk 1 above)
[the remark this answers is not in the document]
Comment (arcondello) [reply dimod/bqm/utils.pyx:46]: This will currently return `sys.maxsize` for an empty BQM, which I think is very confusing behavior. I would rather follow python's behavior and fail when `bqm.num_interactions == 0`, but allow the user to provide a default.

NumPy has a [similar notion](https://numpy.org/doc/stable/reference/generated/numpy.amin.html#numpy.amin).

File: tests/test_minmax.py (hunk 5 above)
Comment (arcondello) [reply tests/test_minmax.py:65]: Also should test with empty BQMs, see note above.

File: tests/test_minmax.py (hunk 4 above)
Comment (arcondello) [reply tests/test_minmax.py:23]: `test_bqm.py` please :smile:

File: dimod/bqm/utils.pyx (hunk 1 above)
[the remark this answers is not in the document]
Comment (randomir) [reply dimod/bqm/utils.pyx:46]: I like NumPy's `initial`.

File: dimod/bqm/utils.pyx [hunk 6]
```diff
@@ -40,6 +42,78 @@ cdef object as_numpy_scalar(double a, np.dtype dtype):
     """Note that the memory is interpreted to match dtype, not a cast"""
     return PyArray_Scalar(&a, dtype, None)
 
+def cylmin(cyBQM bqm, **kwargs):
```
Comment (arcondello) [inline dimod/bqm/utils.pyx:45]: Probably better to name the keyword argument explicitly, e.g `cylmin(cyBQM bqm, object default=None)`. The `object` type specifier is a cython thing, but better explicit than implicit...

One reason to do it this way it that it helps catch hard-to-find errors like
```
cylmin(bqm, dfault=1)
```

File: dimod/bqm/utils.pyx [hunk 7]
```diff
@@ -40,6 +42,78 @@ cdef object as_numpy_scalar(double a, np.dtype dtype):
...
+        else:
+            raise ValueError("Argument is an empty sequence")
+
+    cdef double max_ = -sys.float_info.max
+
+    cdef Py_ssize_t vi
+    for vi in range(bqm.bqm_.num_variables()):
+        span = bqm.bqm_.neighborhood(vi)
+        
+        while span.first != span.second:
```
Comment (arcondello) [inline dimod/bqm/utils.pyx:110]: As a detail, the biases are kept symmetrically, so you could do 
```
while span.first != span.second and deref(span.first).first < vi:
```
which would speed this up a bit.

File: tests/test_bqm.py [hunk 8]
```diff
@@ -2312,3 +2312,49 @@ def test_quadratic_setitem(self, name, BQM):
...
+        bqm = BQM(D, 'SPIN') 
+
+        qmin = min(bqm.quadratic.values())
+        self.assertEqual(qmin, bqm.quadratic.min())
+       
+        qmax = max(bqm.quadratic.values())
+        self.assertEqual(qmax, bqm.quadratic.max())
+
+    @parameterized.expand([(cls.__name__, cls) for cls in BQM_SUBCLASSES])
+    def test_lin_minmax_empty(self, name, BQM):
```
Comment (arcondello) [inline tests/test_bqm.py:2342]: I think this shouldn't be named `test_lin_minmax_empty` since the BQM has 500 variables? I would add some empty tests and make sure that they throw the correct error, something like
```
with self.assertRaises(ValueError):
    BQM('SPIN').linear.min()
self.assertEqual(BQM('SPIN').linear.min(10), 10)
```

File: dimod/bqm/utils.pyx (hunk 6 above)
Comment (arcondello) [reply dimod/bqm/utils.pyx:45]: It would also let people use positional arguments, e.g.
```
cylmin(bqm, 10)
```

**Comment by arcondello** [comment]
Also, these need docstrings. I took a look at how to add them to the documentation, but since they are on the `.linear` and `.quadratic` attributes, it's not so straightforward. If you add the docstrings, I can add them to the docs in a subsequent PR.

File: dimod/bqm/utils.pyx (hunk 6 above)
Comment (randomir) [reply dimod/bqm/utils.pyx:45]: For posterity, the only way to handle the (edge) case of `None` (or any other sentinel) as the actual default value is not to depend on argument value, but existence of argument; therefore the need for `**kwargs`.

But since this is such a rare edge case, I'd agree explicitness trumps it here.

Comment (randomir) [reply dimod/bqm/utils.pyx:45]: FWIW, both unknown kwarg case and `default` as positional argument can be handled "manually."

**Comment by arcondello** [comment]
Closes #402
